dbms/src/AggregateFunctions/AggregateFunctionGroupArray.h

```diff
@@ -43,51 +71,162 @@ struct GroupArrayNumericData
     Array value;
 };
 
+template <typename T>
+struct GroupArrayNumericData<T, true>
+{
+    // Switch to ordinary Allocator after 4096 bytes to avoid fragmentation and trash in Arena
+    using Allocator = MixedAlignedArenaAllocator<alignof(T), 4096>;
+    using Array = PODArray<T, 32, Allocator>;
+
+    Array value;
+    size_t total_values = 0;
+    pcg32_fast rng;
```

Comment:
Separate `rng` for every state looks too much. We can use thread_local_rng.

Reply:
But do we need deterministic result?

Reply:
Perhaps we can have a very unrandom sampler and a truely uniform sampler. 

> But do we need deterministic result?

￼I don't know and I'm always curious about the MergeTree's sample clause and it's deterministic characteristic.

Reply:
`thread_local_rng` will be as random as separate rngs.

> I'm always curious about the MergeTree's sample clause and it's deterministic characteristic.

Yes, but for `groupArraySample` it's not enough to just use deterministic rng, you should also run in single thread to get deterministic result. But it's not practical.